virtcontainers/sandbox.go

```diff
@@ -1898,11 +1898,31 @@ func (s *Sandbox) AddDevice(info config.DeviceInfo) (api.Device, error) {
 	}
 
 	if err := s.devManager.AttachDevice(b.DeviceID(), s); err != nil {
+		if err1 := s.devManager.RemoveDevice(b.DeviceID()); err1 != nil {
+			s.Logger().WithError(err1).
+				WithFields(logrus.Fields{
+					"sandbox":   s.id,
+					"device-ID": b.DeviceID(),
+			    	}).Error("failed to remove device")
+		}
 		return nil, err
 	}
 
 	if !s.supportNewStore() {
 		if err := s.storeSandboxDevices(); err != nil {
+			if err1 := s.devManager.DetachDevice(b.DeviceID(), s); err1 != nil {
```

Comment:
Could you put these in two defer functions?

Reply:
If DetachDevice returns error, I don't think RemoveDevice should be called.

Can you elaborate on how the defer functions should be formed ?

I think the current form is more readable.

Reply:
NewDevice()
defer RemoveDevice()
AttachDevice()
defer DetachDevice()

Reply:
It seems two error variables should be used if we go along the above route.

If DetachDevice call fails, should RemoveDevice be called ?

I think continuing with RemoveDevice call would complicate troubleshooting.